sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4

```diff
@@ -177,6 +177,18 @@ singleTableSchema
     : colTypeList EOF
     ;
 
+singleRoutineParamList
+    : routineParamList EOF
+    ;
+
+routineParamList
+    : routineParam (COMMA routineParam)*
+    ;
+
+routineParam
+    : colType defaultExpression?
```

Comment:
Let's also support COMMENT and NOT NULL here. `NOT NULL` is not really supported and we throw a more user-friendly analysis exception when creating SQL UDFs.

Reply:
+1

Reply:
then how about we support the full col definition syntax and fail for those we don't support as routine parameters? e.g. NOT NULL, GENERATED AS, etc.